### PR TITLE
Added trait printer in device/printer.rs

### PR DESCRIPTION
**Pull request by nicoTolly**
Nearly empty trait, just for starting discussion

### CODE REVIEW

File: src/device/x86/printer.rs [hunk 1]
```diff
@@ -2,517 +2,392 @@ use codegen::*;
...
+    out_function: String,
 }
 
-fn param_decl(param: &ParamVal, namer: &NameMap) -> String {
-    let name = namer.name_param(param.key());
-    match param {
-        ParamVal::External(_, par_type) => format!("{} {}", cpu_type(par_type), name),
-        ParamVal::Size(_) => format!("uint32_t {}", name),
-        ParamVal::GlobalMem(_, _, par_type) => format!("{} {}", cpu_type(par_type), name),
+impl X86printer {
```
Comment (ulysseB) [inline src/device/x86/printer.rs]: You can use `#[derive(Default)]` instead.

File: src/device/x86/printer.rs [hunk 2]
```diff
@@ -2,517 +2,392 @@ use codegen::*;
...
-            assert_eq!(round, op::Rounding::Nearest);
-            let assignement = format!("{} = ",namer.name_inst(inst));
-            format!("{} {} {} {};", assignement, namer.name_op(lhs), binary_op(op), namer.name_op(rhs))
+    /// Declares block and thread indexes.
+    fn decl_par_indexes(&mut self, function: &Function, namer: &mut NameMap) -> String {
+        assert!(function.block_dims().is_empty());
+        let mut decls = vec![];
+        // Compute thread indexes.
+        for (ind, dim) in function.thread_dims().iter().enumerate() {
+            //FIXME: fetch proper thread index
```
Comment (ulysseB) [inline src/device/x86/printer.rs]: Is this FIXME solved ?

File: src/device/x86/printer.rs [hunk 3]
```diff
@@ -2,517 +2,392 @@ use codegen::*;
...
+            // Compute size casts
+            for dim in function.dimensions() {
+                if !dim.kind().intersects(DimKind::UNROLL | DimKind::LOOP) { continue; }
+                for level in dim.induction_levels() {
+                    if let Some((_, incr)) = level.increment {
+                        let name = name_map.declare_size_cast(incr, level.t());
+                        if let Some(name) = name {
+                            let cpu_t = self.get_type(level.t());
+                            let old_name = name_map.name_size(incr, Type::I(32));
+                            self.out_function.push_str(&format!("{} = ({}){};\n", name, cpu_t, old_name));
```
Comment (ulysseB) [inline src/device/x86/printer.rs]: Can't you use `print_cast` instead ?

File: src/device/x86/printer.rs [hunk 4]
```diff
@@ -2,517 +2,392 @@ use codegen::*;
...
+    fn params_call(&mut self, function: &Function) -> String {
+        function.device_code_args()
+            .enumerate().map(|x| x.0)
+            .map(|i| format!("p{}", i))
+            .collect_vec()
+            .join(", ")
+    }
 
+    // Build the right call for a nested loop on dimensions
+    fn build_index_call(&mut self, func: &Function) -> String {
```
Comment (ulysseB) [inline src/device/x86/printer.rs]: I don't understand what this function does.

File: src/device/x86/printer.rs [hunk 5]
```diff
@@ -2,517 +2,392 @@ use codegen::*;
...
-}
+        let mut ret = format!("pthread_t thr_ids[{}];\n", func.num_threads());
+        let mut ind_var_decl = String::from("int ");
+        let build_struct = format!("thread_arg_t thread_args[{}];\n", func.num_threads());
+        let dim_tid_struct = format!("thread_dim_id_t thread_tids[{}];\n", func.num_threads());
+        let barrier_init = format!("pthread_barrier_init(&barrier, NULL,{});\n",   func.num_threads() );
+        let mut loop_decl = String::new();
+        let mut ind_vec = Vec::new();
+        let mut jmp_stack = Vec::new();
+        for (ind, dim) in func.thread_dims().iter().enumerate() {
```
Comment (ulysseB) [inline src/device/x86/printer.rs]: maybe use a template instead ?

File: src/device/x86/printer.rs [hunk 6]
```diff
@@ -2,517 +2,392 @@ use codegen::*;
...
-            standard_loop(fun, dim, cfgs, namer)
+    fn thread_join(&mut self, func: &Function) -> String {
+        if func.num_threads() == 1 {
+            return String::new();
+        }
+        let mut ret = String::new();
+        let mut loop_decl = String::new();
+        let mut jmp_stack = Vec::new();
+        for (ind, dim) in func.thread_dims().iter().enumerate() {
+            let mut loop_jmp = String::new();
```
Comment (ulysseB) [inline src/device/x86/printer.rs]: Maybe use a template instead ?

File: src/codegen/printer.rs [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+    /// Change the side-effect guards so that only the specified threads are enabled.
+    fn enable_threads(&mut self, fun: &Function, threads: &[bool], namer: &mut NameMap) {
+        let mut guard = None;
+        for (&is_active, dim) in threads.iter().zip_eq(fun.thread_dims().iter()) {
+            if is_active { continue; }
+            let new_guard = namer.gen_name(ir::Type::I(1));
+            let index = namer.name_index(dim.id());
+            self.print_equal(&new_guard, index, &"0");
+            if let Some(ref guard) = guard {
+                self.print_and(&guard as &str, &guard as &str, &new_guard);
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done (needs to explicitly tell the type to the compiler)

File: src/codegen/printer.rs [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    }
+
+    /// Change the side-effect guards so that only the specified threads are enabled.
+    fn enable_threads(&mut self, fun: &Function, threads: &[bool], namer: &mut NameMap) {
+        let mut guard = None;
+        for (&is_active, dim) in threads.iter().zip_eq(fun.thread_dims().iter()) {
+            if is_active { continue; }
+            let new_guard = namer.gen_name(ir::Type::I(1));
+            let index = namer.name_index(dim.id());
+            self.print_equal(&new_guard, index, &"0");
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/codegen/printer.rs [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+    fn parallel_induction_level(&mut self, level: &InductionLevel, namer: &NameMap) {
+        let dim_id = level.increment.map(|(dim, _)| dim);
+        let ind_var = namer.name_induction_var(level.ind_var, dim_id);
+        let base_components =  level.base.components().map(|v| namer.name(v)).collect_vec();
+        if let Some((dim, increment)) = level.increment {
+            let index = namer.name_index(dim);
+            let step = namer.name_size(increment, Type::I(32));
+            let mode = if level.t() == Type::I(64) { MulMode::Wide } else { MulMode::Low };
+            match base_components[..] {
+                // TODO get a sensible type for index and step rather than just random I32
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/codegen/printer.rs [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+    /// Prints a multiplicative induction var level.
+    fn parallel_induction_level(&mut self, level: &InductionLevel, namer: &NameMap) {
+        let dim_id = level.increment.map(|(dim, _)| dim);
+        let ind_var = namer.name_induction_var(level.ind_var, dim_id);
+        let base_components =  level.base.components().map(|v| namer.name(v)).collect_vec();
+        if let Some((dim, increment)) = level.increment {
+            let index = namer.name_index(dim);
+            let step = namer.name_size(increment, Type::I(32));
+            let mode = if level.t() == Type::I(64) { MulMode::Wide } else { MulMode::Low };
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/codegen/printer.rs [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+    /// Print if (cond) jump label(label_id)
+    fn print_cond_jump(&mut self, label_id: &str, cond: &str);
+
+    /// Print wait on all threads
+    fn print_sync(&mut self);
+
+    fn mul_mode(from: Type, to: Type) -> MulMode {
+        match (from, to) {
+            (Type::I(32), Type::I(64)) => MulMode::Wide,
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/device/cuda/gpu.rs [hunk 12]
```diff
@@ -321,7 +322,7 @@ impl Gpu {
 }
 
 impl device::Device for Gpu {
-    fn print(&self, fun: &Function, out: &mut Write) { p::host_function(fun, self, out) }
+    fn print(&self, fun: &Function, out: &mut Write) { let mut printer = CudaPrinter::new(); printer.host_function(fun, self, out) }
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/device/cuda/gpu.rs]: Done

File: src/codegen/printer.rs [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+        let loop_id = namer.gen_loop_id();
+        let ind_levels = dim.induction_levels();
+        for level in ind_levels.iter() {
+            let dim_id = level.increment.map(|(dim, _)| dim);
+            let ind_var = namer.name_induction_var(level.ind_var, dim_id);
+            let base_components = level.base.components().map(|v| namer.name(v));
+            match base_components.collect_vec()[..] {
+                [ref base] => self.print_mov(level.t(), &ind_var, &base),
+                [ref lhs, ref rhs] =>
+                    self.print_binop( ir::BinOp::Add, level.t(), op::Rounding::Exact, &ind_var, &lhs, &rhs, ),
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/device/x86/printer.rs (hunk 1 above)
Comment (nicoTolly) [reply src/device/x86/printer.rs]: Done

File: src/device/x86/printer.rs (hunk 2 above)
Comment (nicoTolly) [reply src/device/x86/printer.rs]: Done

File: src/device/x86/printer.rs (hunk 4 above)
Comment (nicoTolly) [reply src/device/x86/printer.rs]: Hope it's better now

File: src/device/x86/printer.rs (hunk 5 above)
Comment (nicoTolly) [reply src/device/x86/printer.rs]: Tried to clean up that

File: src/device/x86/printer.rs (hunk 6 above)
Comment (nicoTolly) [reply src/device/x86/printer.rs]: Same

File: src/device/cuda/printer.rs [hunk 14]
```diff
@@ -1,553 +1,264 @@
...
+    /// Prints a store operator.
+    fn st_operator(flag: InstFlag) -> &'static str {
+        match flag {
+            InstFlag::MEM_SHARED => "st.shared",
+            InstFlag::MEM_CA => "st.global.wb",
+            InstFlag::MEM_CG => "st.global.cg",
+            InstFlag::MEM_CS => "st.global.cs",
+            _ => panic!("invalid store flag {:?}", flag),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/device/cuda/printer.rs]: Done

File: src/codegen/printer.rs [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+                let low_mlhs_type = Self::lower_type(mul_lhs.t(), fun);
+                let low_arhs_type = Self::lower_type(add_rhs.t(), fun);
+                let mode = if round == op::Rounding::Exact {
+                    Self::mul_mode(low_mlhs_type, low_arhs_type)
+                } else {MulMode::Empty };
+                self.print_mad(Self::lower_type(inst.t(), fun), round, mode, &namer.name_inst(inst), &namer.name_op(mul_lhs), &namer.name_op(mul_rhs), &namer.name_op(add_rhs))
+            },
+            op::Mov(ref op) => {
+
+                self.print_mov(Self::lower_type(inst.t(), fun), &namer.name_inst(inst), &namer.name_op(op))
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/codegen/printer.rs [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+                } else {MulMode::Empty };
+                self.print_mul(low_ret_type, round, mode, &namer.name_inst(inst), &namer.name_op(lhs), &namer.name_op(rhs))
+            },
+            op::Mad(ref mul_lhs, ref mul_rhs, ref add_rhs, round) => {
+                let low_mlhs_type = Self::lower_type(mul_lhs.t(), fun);
+                let low_arhs_type = Self::lower_type(add_rhs.t(), fun);
+                let mode = if round == op::Rounding::Exact {
+                    Self::mul_mode(low_mlhs_type, low_arhs_type)
+                } else {MulMode::Empty };
+                self.print_mad(Self::lower_type(inst.t(), fun), round, mode, &namer.name_inst(inst), &namer.name_op(mul_lhs), &namer.name_op(mul_rhs), &namer.name_op(add_rhs))
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Ok

File: src/codegen/printer.rs [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+    /// Prints an instruction.
+    fn inst(&mut self, inst: &Instruction, namer: &mut NameMap, fun: &Function ) {
+        match *inst.operator() {
+            op::BinOp(op, ref lhs, ref rhs, round) => {
+                self.print_binop(op, inst.t(), round, &namer.name_inst(inst), &namer.name_op(lhs), &namer.name_op(rhs))
+            }
+            op::Mul(ref lhs, ref rhs, round, return_type) => {
+                let low_lhs_type = Self::lower_type(lhs.t(), fun);
+                let low_ret_type = Self::lower_type(return_type, fun);
+                let mode = if round == op::Rounding::Exact {
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/codegen/printer.rs [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+    fn gen_loop(&mut self, fun: &Function, dim: &Dimension, cfgs:
+                                &[Cfg], namer: &mut NameMap)
+    {
+        match dim.kind() {
+            DimKind::LOOP => {
+                self.standard_loop(fun, dim, cfgs, namer)
+            }
+            DimKind::UNROLL => {self.unroll_loop(fun, dim, cfgs, namer)}
+            DimKind::VECTOR => {self.unroll_loop(fun, dim, cfgs, namer)}
+            _ => { unimplemented!() }
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/device/x86/printer.rs (hunk 3 above)
Comment (nicoTolly) [reply src/device/x86/printer.rs]: Done

File: src/codegen/printer.rs [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+        self.cfg_vec(fun, cfgs, namer);
+        for (level, ind_var) in ind_levels.iter().zip_eq(ind_var_vec) {
+            if let Some((_, increment)) = level.increment {
+                let step = namer.name_size(increment, level.t());
+                self.print_binop(ir::BinOp::Add, level.t(), op::Rounding::Exact, &ind_var, &ind_var, &step);
+            };
+        }
+        let one = self.get_int(1);
+        self.print_binop(op::BinOp::Add, Type::I(32), op::Rounding::Exact, &idx, &idx,  &one);
+        let gt_cond = namer.gen_name(ir::Type::I(1));
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

File: src/codegen/printer.rs [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+        self.print_lt(&gt_cond, &idx, &namer.name_size(dim.size(), Type::I(32)));
+        self.print_cond_jump(&loop_id.to_string(), &gt_cond);
+    }
+
+    fn unroll_loop(&mut self, fun: &Function, dim: &Dimension, cfgs: &[Cfg], namer: &mut NameMap) {
+        //let mut body = Vec::new();
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/printer.rs]: Done

**Pull request merged**